Content.Server/Traits/TraitSystem.Functions.cs

```diff
@@ -0,0 +1,280 @@
+using Robust.Shared.Serialization.TypeSerializers.Implementations.Custom.Prototype.Set;
+using Content.Shared.Actions;
+using Content.Server.Abilities.Psionics;
+using Content.Shared.Psionics;
+using Content.Server.Language;
+using Content.Shared.Mood;
+using Content.Server.NPC.Systems;
+using Robust.Shared.Serialization;
+
+namespace Content.Server.Traits;
+
+/// Used for traits that add a Component upon spawning in, overwriting the pre-existing component if it already exists.
+[UsedImplicitly]
+public sealed partial class TraitReplaceComponent : TraitFunction
+{
+    [DataField, AlwaysPushInheritance]
+    public ComponentRegistry Components { get; private set; } = new();
+
+    public override void OnPlayerSpawn(EntityUid uid,
+        IComponentFactory factory,
+        IEntityManager entityManager,
+        ISerializationManager serializationManager)
+    {
+        foreach (var (name, data) in Components)
+        {
+            var component = (Component) factory.GetComponent(name);
+            component.Owner = uid;
+
+            var temp = (object) component;
+            serializationManager.CopyTo(data.Component, ref temp);
+            entityManager.RemoveComponent(uid, temp!.GetType());
+            entityManager.AddComponent(uid, (Component) temp);
+        }
+    }
+}
+
+/// <summary>
+///     Used for traits that add a Component upon spawning in.
+///     This will do nothing if the Component already exists.
+/// </summary>
+[UsedImplicitly]
+public sealed partial class TraitAddComponent : TraitFunction
+{
+    [DataField, AlwaysPushInheritance]
+    public ComponentRegistry Components { get; private set; } = new();
+
+    public override void OnPlayerSpawn(EntityUid uid,
+        IComponentFactory factory,
+        IEntityManager entityManager,
+        ISerializationManager serializationManager)
+    {
+        foreach (var (name, _) in Components)
+        {
+            var component = (Component) factory.GetComponent(name);
+            component.Owner = uid;
+
+            entityManager.AddComponent(uid, component);
+        }
+    }
+}
+
+/// Used for traits that remove a component upon a player spawning in.
+[UsedImplicitly]
+public sealed partial class TraitRemoveComponent : TraitFunction
+{
+    [DataField, AlwaysPushInheritance]
+    public ComponentRegistry Components { get; private set; } = new();
+
+    public override void OnPlayerSpawn(EntityUid uid,
+        IComponentFactory factory,
+        IEntityManager entityManager,
+        ISerializationManager serializationManager)
+    {
+        foreach (var (name, _) in Components)
+            entityManager.RemoveComponent(uid, (Component) factory.GetComponent(name));
+    }
+}
+
+/// Used for traits that add an action upon a player spawning in.
+[UsedImplicitly]
+public sealed partial class TraitAddActions : TraitFunction
+{
+    [DataField, AlwaysPushInheritance]
+    public List<EntProtoId>? Actions { get; private set; } = default!;
```

Comment:
in no world this should be nullable.
```suggestion
    [DataField(required: true), AlwaysPushInheritance]
    public List<EntProtoId> Actions { get; private set; } = new();
```